sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/StringFormatValidatorBase.cs

```diff
@@ -0,0 +1,444 @@
+                !currentMethod.Name.Contains("Format"));
+        }
+
+        private static ValidationFailure CheckForIssues(string formatStringText, ArgumentListSyntax argumentList,
+            int formatArgumentIndex, SemanticModel semanticModel)
+        {
+            if (formatStringText == null)
+            {
+                return new ValidationFailure(ValidationFailure.FailureKind.NullFormatString);
+            }
+
+            List<FormatStringItem> formatStringItems;
+            return ExtractFormatItems(formatStringText, out formatStringItems) ??
+                TryValidateFormatString(formatStringText, formatStringItems, argumentList, formatArgumentIndex,
+                    semanticModel);
+        }
+
+        private static ValidationFailure ExtractFormatItems(string formatString,
+            out List<FormatStringItem> formatStringItems)
+        {
+            formatStringItems = new List<FormatStringItem>();
+            var curlyBraceCount = 0;
+            StringBuilder currentFormatItemBuilder = null;
+            var isEscapingOpenCurlyBrace = false;
+            var isEscapingCloseCurlyBrace = false;
+            for (int i = 0; i < formatString.Length; i++)
+            {
+                var currentChar = formatString[i];
+                var previousChar = i > 0 ? formatString[i - 1] : '\0';
+
+                if (currentChar == '{')
+                {
+                    if (previousChar == '{' && !isEscapingOpenCurlyBrace)
+                    {
+                        curlyBraceCount--;
+                        isEscapingOpenCurlyBrace = true;
+                        currentFormatItemBuilder = null;
+                        continue;
+                    }
+
+                    curlyBraceCount++;
+                    isEscapingOpenCurlyBrace = false;
+                    if (currentFormatItemBuilder == null)
+                    {
+                        currentFormatItemBuilder = new StringBuilder();
+                    }
+                    continue;
+                }
+
+                if (previousChar == '{' && !char.IsDigit(currentChar) && currentFormatItemBuilder != null)
+                {
+                    return new ValidationFailure(ValidationFailure.FailureKind.InvalidCharacterAfterOpenCurlyBrace);
+                }
+
+                if (currentChar == '}')
+                {
+                    isEscapingCloseCurlyBrace = previousChar == '}' && !isEscapingCloseCurlyBrace;
+                    curlyBraceCount = isEscapingCloseCurlyBrace
+                        ? curlyBraceCount + 1
+                        : curlyBraceCount - 1;
+
+                    if (currentFormatItemBuilder != null)
+                    {
+                        FormatStringItem formatStringItem;
+                        var failure = TryParseItem(currentFormatItemBuilder.ToString(), out formatStringItem);
+                        if (failure != null)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Refactor this code to not nest more than 3 control flow statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS134)